### DISCUSSION TITLE
How to open/close Accordion programmatically

### DISCUSSION THREAD

**Discussion by irhamputra** [radix-ui/primitives#824] [Help] [answered]
Hi Author and the community,

I'm asking for a help about how to open/close the accordion programmatically? I've been searching in the docs and the issue there is no answer related to this topic.

Any ideas and workaround would be appreciated.
Thank you

**Comment by jjenzz** [comment, marked as answer]
All of our components can be "controlled" if you would like to open/close them programmatically.

```jsx
const [value, setValue] = React.useState('one');

return () => (
  <div>
	{/* Something outside of Accordion that controls the open content */}
    <button onClick={() => setValue('two')}>Open two</button>

    <Accordion.Root type="single" value={value} onValueChange={setValue}>
      <Accordion.Item value="one">
        <Accordion.Header>
          <Accordion.Trigger>One</Accordion.Trigger>
        </Accordion.Header>
        <Accordion.Content>
          Content for one
        <Accordion.Content>
      </Accordion.Item>
      <Accordion.Item value="two">
        <Accordion.Header>
          <Accordion.Trigger>Two</Accordion.Trigger>
        </Accordion.Header>
        <Accordion.Content>
          Content for two
        <Accordion.Content>
      </Accordion.Item>
    </Accordion.Root>
  </div>
```

**Reply by irhamputra** [reply]
thanks a lot

**Reply by yalamber** [reply]
how to open multiple items at once?

**Reply by benoitgrelard** [reply]
> how to open multiple items at once?

Exactly the same but when `type="multiple"` the state is an array instead.

**Reply by sunilmauyra** [reply]
thanks you very much,

**Reply by manavm1990** [reply]
Copy and paste this simple example into the docs 📝 ❓ @jjenzz Respectfully, there's nothing major here, and yet it just didn't occur to me to do this until I took 5 seconds to look at the code!

**Reply by obrunofontana** [reply]
its work. thank you!!

**Reply by TomasHubelbauer** [reply]
Anyone else able to make this work to close the opened `single` `AccordionItem`? When I set the state driving `value` to `undefined`, the currently opened item just remains opened. :/

**Reply by manavm1990** [reply]
Maybe 🤔 you can share some code? You may not need even need any `State` if it's `"single"`.

**Reply by jjenzz** [reply]
@TomasHubelbauer are you looking for the `collapsible` prop by any chance?

![Screenshot 2024-03-31 at 17 03 28](https://github.com/radix-ui/primitives/assets/175330/439ec477-541a-47f9-98ef-c236779d1b23)

**Reply by eliuAtFanatics** [reply]
> Anyone else able to make this work to close the opened `single` `AccordionItem`? When I set the state driving `value` to `undefined`, the currently opened item just remains opened. :/

I'm seeing this same behavior as well right now with dynamically programming the value.  Has anyone been able to fix this when the value is set to undefined?

**Reply by thanksphil** [reply]
> Anyone else able to make this work to close the opened `single` `AccordionItem`? When I set the state driving `value` to `undefined`, the currently opened item just remains opened. :/

Also experiencing this issue. I have seen this with both a regular state hook and when using mobx-state-tree state; when the state (`value`) is set to `undefined` is doesn't collapse the accordion item, though for me clicking a second time does then collapse it.

**Reply by charpeni** [reply]
I experienced the same issue as described above where a controlled `<Accordion.Root>` doesn't allow being closed by passing `undefined` as a value. I ended up setting the value to an empty string: `setValue('')`. This closes it probably because it triggers a change combined with `collapsible` and `single`, which doesn't allow multiple accordions to be open, but the newly set value doesn't match with any existing accordions.

cc @TomasHubelbauer @eliuAtFanatics @thanksphil

**Reply by poulch** [reply]
> I experienced the same issue as described above where a controlled `<Accordion.Root>` doesn't allow being closed by passing `undefined` as a value. I ended up setting the value to an empty string: `setValue('')`. This closes it probably because it triggers a change combined with `collapsible` and `single`, which doesn't allow multiple accordions to be open, but the newly set value doesn't match with any existing accordions.
> 
> cc @TomasHubelbauer @eliuAtFanatics @thanksphil

It works, thanks @charpeni

**Reply by benjick** [reply]
For single accordion:

```tsx
function MyComponent() {
  const [isOpen, setIsOpen] = useState(true);

  return (
    <Accordion
      type="single"
      collapsible
      value={isOpen ? "open" : "closed"}
      onValueChange={(value) => setIsOpen(value === "open")}
    >
      <AccordionItem value="open" className="border rounded-lg">
      ...
      </AccordionItem>
    </Accordion>
  );
}
```

**Comment by KarthikRaju391** [comment]
Is there a way to know the open or closed state of an accordion?

**Reply by joaom00** [reply]
via css you can use the `data-state="open | closed"` attribute to style and via js controlling the component like https://github.com/radix-ui/primitives/discussions/824#discussioncomment-1194231

**Reply by emilioaray-dev** [reply]
If possible, here's an example evaluating the [data-state] value and using useRef in case you need to map your component.
```
import React, { useEffect, useRef, useState } from "react";
import { Accordion, AccordionTrigger, AccordionContent, AccordionItem } from "@radix-ui/react-accordion";

const YourComponent = () => {
  const [isOpenAccordion, setIsOpenAccordion] = useState(null);
  const [isClicked, setIsClicked] = useState<boolean | null>(null);
  const accordionTriggerRef = useRef(null);

  useEffect(() => {
    if (accordionTriggerRef.current) {
      const accordionTriggerElement = accordionTriggerRef.current;
      const dataState = accordionTriggerElement.getAttribute("data-state");

      if (dataState === "closed") {
        setIsOpenAccordion(false);
        console.log("AccordionTrigger is closed.");
      }

      if (dataState === "open") {
        setIsOpenAccordion(true);
        console.log("AccordionTrigger is open.");
      }
    }
  }, [isClicked]);

  return (
    <>
      <div>
        <Accordion type="multiple">
          <AccordionItem ref={accordionTriggerRef} value="YourAccordion">
            <AccordionTrigger
              onClick={() => {
                setIsClicked(!isClicked);
              }}
            >
              {/* Your AccordionTrigger UI here */}
            </AccordionTrigger>
            <AccordionContent>
              {isOpenAccordion ? (
                // Content to display when the Accordion is open
                <p>Accordion Content (This content is shown only if the Accordion is open)</p>
              ) : null}
            </AccordionContent>
          </AccordionItem>
        </Accordion>
      </div>
    </>
  );
};

export default YourComponent;
```

I hope you find it useful. 👍

**Reply by jjenzz** [reply]
@emilioaray-dev i'm a bit late seeing this but as @joaom00 suggested, the controlled API is meant for this and it would make your code simpler (more declarative):

```tsx
import React, { useEffect, useRef, useState } from "react";
import { Accordion, AccordionTrigger, AccordionContent, AccordionItem } from "@radix-ui/react-accordion";

const YourComponent = () => {
  const [values, setValues] = React.useState<string[]>([]);

  return (
    <>
      <div>
        <Accordion type="multiple" value={values} onValueChange={setValues}>
          <AccordionItem value="YourAccordion">
            <AccordionTrigger>
              {/* Your AccordionTrigger UI here */}
            </AccordionTrigger>
            <AccordionContent>
              {/* derive the `isOpen` state based on `values` */}
              {values.includes("YourAccordion") ? (
                // Content to display when the Accordion is open
                <p>Accordion Content (This content is shown only if the Accordion is open)</p>
              ) : null}
            </AccordionContent>
          </AccordionItem>
        </Accordion>
      </div>
    </>
  );
};

export default YourComponent;
```

**Reply by jewel109** [reply]
@jjenzz how can i trigger <AccordionTrigger> from other component? suppose I will click a button which is not part of the accordion component.

**update**:
I solved it from Your first answer.

**Comment by sigmarsson** [comment]
Could anyone explain me what the purpose is of `<AccordionTrigger>` instead of `<Accordion.Trigger>` ?

**Reply by jjenzz** [reply]
they are the same. both parts exist because people like to import in different ways depending on the tree-shaking capabilities of their bundler. the convention is `import * as Accordion` (so you can use `Accordion.Trigger`) but some bundlers might not tree-shake appropriately there. for that case, you can `import { AccordionTrigger }`.

**Reply by sigmarsson** [reply]
Ahh, merci! - I have not been thinking from that perspective.

**Comment by TobiasLueger** [comment]
Hey all,
I followed this conversation. How can I open all ans close all with two seperate Buttons. I tried something like that but it doesn't work.

```typescript
import { Root as RDXAccordion } from "@radix-ui/react-accordion";
import clsx from "clsx";
import { Children, isValidElement, useEffect, useState } from "react";

import styles from "./Accordion.module.scss";
import { AccordionProps } from "./types.ts";

export function Accordion({
  children,
  className,
  type,
  dataTestId,
  openAll,
  ...props
}: AccordionProps) {
  const [openAccordionItems, setOpenAccordionItems] = useState<string[]>();

  const setOpenValues = (values: string[]) => {
    setOpenAccordionItems(values);
  };

  useEffect(() => {
    const convertAllAccordionItemValuesToStringArray = () => {
      const values: string[] = [];
      const getAllAccordionItemValues = (children: React.ReactNode) => {
        Children.forEach(children, (child) => {
          if (isValidElement(child)) {
            console.log("Test", child);
            if ((child.type as any).displayName === "AccordionItem")
              return values.push((child as React.ReactElement).props.value);
            return getAllAccordionItemValues(
              (child as React.ReactElement).props.children
            );
          }
        });
      };
      getAllAccordionItemValues(children);
      return values;
    };

    if (openAll) {
      setOpenAccordionItems(convertAllAccordionItemValuesToStringArray());
    } else {
      setOpenAccordionItems([]);
    }
  }, [children, openAll]);

  return (
    <RDXAccordion
      type={type || "multiple"}
      className={clsx(className, styles.accordion)}
      data-testid={dataTestId || "accordion"}
      value={openAccordionItems} //there has to be a array of AccordionItem Values to open all accordions
      onValueChange={setOpenValues}
      {...props}
    >
      {children}
    </RDXAccordion>
  );
}`
``` 

I hope someone have an Idea 🔥 ❤️

**Comment by uzzal50** [comment]
@jjenzz  Hello jenna I want my accrdion item to disable when its data state is open how can i achieve this thank you ?
